File: gemm/OpenMP.f90

program main
    implicit none
    integer, parameter :: m = 3, n = 3, k = 3, lda = 3, ldb = 3, ldc = 3
    real(kind=8), dimension(lda, k) :: a
    real(kind=8), dimension(ldb, n) :: b
    real(kind=8), dimension(ldc, n) :: c
    real(kind=8) :: alpha, beta
    integer :: i

    ! Initialize alpha, beta, a, b, and c here...
    alpha = 1.0
    beta = 0.0
    a = reshape([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0], [lda, k])
    b = reshape([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0], [ldb, n])
    c = 0.0

    call gemm_parallel(m, n, k, alpha, a, lda, b, ldb, beta, c, ldc)

    ! Output the resulting matrix C
    print *, 'Resulting matrix C:'
    do i = 1, m
      print '(3F8.2)', C(i, :)
    end do
end program main

subroutine gemm_parallel(m, n, k, alpha, a, lda, b, ldb, beta, c, ldc)
    implicit none
    integer, intent(in) :: m, n, k, lda, ldb, ldc
    real(kind=8), intent(in) :: alpha, beta
    real(kind=8), intent(in) :: a(lda, k), b(ldb, n)
    real(kind=8), intent(inout) :: c(ldc, n)
    integer :: i, j, l
    real(kind=8) :: temp
  
    !$OMP PARALLEL DO PRIVATE(i, j, l, temp) SCHEDULE(STATIC)
    do j = 1, n
      do i = 1, m
        temp = 0.0
        do l = 1, k
          temp = temp + a(i, l) * b(l, j)
        end do
        c(i, j) = alpha * temp + beta * c(i, j)
      end do
    end do
    !$OMP END PARALLEL DO
  end subroutine gemm_parallel

File: gemm/performance/OpenMP.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <time.h>  // for timing

void gemm_parallel(int m, 
                   int n, 
                   int k, 
                   double alpha, 
                   const double *a, 
                   int lda, 
                   const double *b, 
                   int ldb, 
                   double beta, 
                   double *c, 
                   int ldc) 
{
    int i, j, l;
    double temp;

    #pragma omp parallel for private(i, j, l, temp) schedule(static)
    for (j = 0; j < n; j++) {
        for (i = 0; i < m; i++) {
            temp = 0.0;
            for (l = 0; l < k; l++) {
                temp += a[i * lda + l] * b[l * ldb + j];
            }
            c[i * ldc + j] = alpha * temp + beta * c[i * ldc + j];
        }
    }
}

int main() {
    int m = 25027, n = 25027, k = 25027;
    double alpha = 1.0, beta = 1.0;
    double *a = (double *)malloc(m * k * sizeof(double));
    double *b = (double *)malloc(k * n * sizeof(double));
    double *c = (double *)malloc(m * n * sizeof(double));

    omp_set_num_threads(64);

    int maxThreads = omp_get_max_threads();
    printf("Threads set to: %d\n", maxThreads);

    // Initialize matrices A and B
    for (int i = 0; i < m * k; i++) {
        a[i] = (double)(i % 100) / 100.0;
    }
    for (int i = 0; i < k * n; i++) {
        b[i] = (double)(i % 100) / 100.0;
    }

    // Warm-up
    gemm_parallel(m, n, k, alpha, a, k, b, n, beta, c, n);

    // Timing 10 runs and calculating average time
    clock_t start, end;
    double total_time = 0.0;
    int runs = 10;

    for (int i = 0; i < runs; i++) {
    double start_time = omp_get_wtime();
    gemm_parallel(m, n, k, alpha, a, k, b, n, beta, c, n);
    double end_time = omp_get_wtime();
    double elapsed_time = end_time - start_time;
    total_time += elapsed_time;
    printf("Run %d: %f seconds\n", i, elapsed_time);
}

double average_time = total_time / runs;
printf("Average time taken for %d runs: %f seconds\n", runs, average_time);

free(a);
free(b);
free(c);

return 0;
}